File: sdk/go/autonomy.go
package alpgo

import (
	"fmt"
	"time"
)

type EditProposal struct {
	ProposalID  string
	WorkflowID  string
	Edits       []map[string]any
	Rationale   string
	Status      string
	CreatedAt   string
	ReviewedAt  *string
	ReviewNote  *string
}

func NewEditProposal(workflowID string, edits []map[string]any, rationale string) *EditProposal {
	proposalID := fmt.Sprintf("prop-%s-%d", workflowID, len(edits)+1)
	createdAt := time.Now().Format(time.RFC3339)
	return &EditProposal{
		ProposalID: proposalID,
		WorkflowID: workflowID,
		Edits:      edits,
		Rationale:  rationale,
		Status:     "pending",
		CreatedAt:  createdAt,
	}
}

func (p *EditProposal) ToDict() map[string]any {
	dict := map[string]any{
		"proposal_id": p.ProposalID,
		"workflow_id": p.WorkflowID,
		"edits":       p.Edits,
		"rationale":   p.Rationale,
		"status":      p.Status,
		"created_at":  p.CreatedAt,
	}
	if p.ReviewedAt != nil {
		dict["reviewed_at"] = *p.ReviewedAt
	}
	if p.ReviewNote != nil {
		dict["review_note"] = *p.ReviewNote
	}
	return dict
}

type WorkflowMutator struct {
	policyEngine     *PolicyEngine
	proposals        map[string]*EditProposal
	rollbackSnapshots map[string]map[string]any
}

func NewWorkflowMutator() *WorkflowMutator {
	return &WorkflowMutator{
		proposals:        make(map[string]*EditProposal),
		rollbackSnapshots: make(map[string]map[string]any),
	}
}

func (w *WorkflowMutator) WithPolicyEngine(engine *PolicyEngine) *WorkflowMutator {
	w.policyEngine = engine
	return w
}

func (w *WorkflowMutator) ProposeEdit(workflowID string, edits []map[string]any, rationale string) *EditProposal {
	proposal := NewEditProposal(workflowID, edits, rationale)
	w.proposals[proposal.ProposalID] = proposal
	return proposal
}

func (w *WorkflowMutator) Approve(proposalID string, workflow map[string]any) (map[string]any, error) {
	proposal, ok := w.proposals[proposalID]
	if !ok {
		return nil, fmt.Errorf("proposal %s not found", proposalID)
	}
	if w.policyEngine != nil {
		_ = w.policyEngine.Evaluate(NewPolicyQuery("edits", proposalID))
	}
	w.rollbackSnapshots[proposalID] = workflow
	updated := workflow
	for _, edit := range proposal.Edits {
		for k, v := range edit {
			updated[k] = v
		}
	}
	proposal.Status = "approved"
	reviewedAt := time.Now().Format(time.RFC3339)
	proposal.ReviewedAt = &reviewedAt
	return updated, nil
}

func (w *WorkflowMutator) Rollback(proposalID string) map[string]any {
	snapshot, ok := w.rollbackSnapshots[proposalID]
	if ok {
		delete(w.rollbackSnapshots, proposalID)
	}
	if proposal, ok := w.proposals[proposalID]; ok {
		proposal.Status = "rolled_back"
	}
	return snapshot
}

func (w *WorkflowMutator) GetProposal(proposalID string) *EditProposal {
	return w.proposals[proposalID]
}

func (w *WorkflowMutator) ListProposals() []*EditProposal {
	result := make([]*EditProposal, 0, len(w.proposals))
	for _, p := range w.proposals {
		result = append(result, p)
	}
	return result
}

File: sdk/rust/src/autonomy.rs
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct EditProposal {
    pub proposal_id: String,
    pub workflow_id: String,
    pub edits: Vec<HashMap<String, serde_json::Value>>,
    pub rationale: String,
    pub status: String,
    pub created_at: String,
    pub reviewed_at: Option<String>,
    pub review_note: Option<String>,
}

#[derive(Debug, Clone)]
pub struct WorkflowMutator {
    pub policy_engine: Option<crate::PolicyEngine>,
    proposals: HashMap<String, EditProposal>,
    rollback_snapshots: HashMap<String, HashMap<String, serde_json::Value>>,
}

impl WorkflowMutator {
    pub fn new() -> Self {
        Self {
            policy_engine: None,
            proposals: HashMap::new(),
            rollback_snapshots: HashMap::new(),
        }
    }

    pub fn with_policy_engine(mut self, engine: crate::PolicyEngine) -> Self {
        self.policy_engine = Some(engine);
        self
    }

    pub fn propose_edit(&mut self, workflow_id: impl Into<String>, edits: Vec<HashMap<String, serde_json::Value>>, rationale: impl Into<String>) -> EditProposal {
        let proposal_id = format!("prop-{}-{}", workflow_id.into(), self.proposals.len() + 1);
        let proposal = EditProposal {
            proposal_id: proposal_id.clone(),
            workflow_id: workflow_id.into(),
            edits,
            rationale: rationale.into(),
            status: "pending".into(),
            created_at: chrono::Utc::now().to_rfc3339(),
            reviewed_at: None,
            review_note: None,
        };
        self.proposals.insert(proposal_id.clone(), proposal.clone());
        proposal
    }

    pub fn approve(&mut self, proposal_id: &str, workflow: &HashMap<String, serde_json::Value>) -> Result<HashMap<String, serde_json::Value>, AlpError> {
        let proposal = self.proposals.get(proposal_id).ok_or_else(|| AlpError::new(format!("Proposal {} not found.", proposal_id)))?;
        if let Some(ref engine) = self.policy_engine {
            let _ = engine.evaluate(&crate::PolicyQuery::new("edits", proposal_id));
        }
        self.rollback_snapshots.insert(proposal_id.into(), workflow.clone());
        let mut updated = workflow.clone();
        for edit in &proposal.edits {
            for (key, value) in edit {
                updated.insert(key.clone(), value.clone());
            }
        }
        Ok(updated)
    }

    pub fn rollback(&mut self, proposal_id: &str) -> Option<HashMap<String, serde_json::Value>> {
        let snapshot = self.rollback_snapshots.remove(proposal_id);
        if let Some(mut proposal) = self.proposals.get_mut(proposal_id) {
            proposal.status = "rolled_back".into();
        }
        snapshot
    }
}
